### PR TITLE
SymbolGraph: Serialize source locations and doc comment ranges

### PR DESCRIPTION
**Pull request by bitjammer**
- Add DocRangesLayout to the `.swiftsourceinfo`.
  This is a blob containing an array of `SingleRawComment`
  source locations.
    
- Add DocLocWriter for serializing `SingleRawComment` locs into the
  `DocLocsLayout` buffer.
  Serialize start line, start column, and length of `SingleRawComment`
  pieces in `.swiftsourceinfo`
    
- Read doc locs when loading basic declaration locs from a ModuleFile.
  - Load `DOC_LOCS` blob into ModuleFile::DocLocsData
  - Reconstitute RawComment ranges when available from .swiftsourceinfo
    
- Allow requesting serialized raw comment if available

rdar://problem/58339492

### CODE REVIEW

**Comment by bitjammer** [comment]
@swift-ci Please smoke test

**Comment by bitjammer** [comment]
@swift-ci Please test Windows platform

**Comment by bitjammer** [comment]
@swift-ci please smoke test compiler performance

**Comment by bitjammer** [comment]
@swift-ci Please smoke test

**Comment by bitjammer** [comment]
@swift-ci Please test Windows platform

**Comment by bitjammer** [comment]
@swift-ci please smoke test compiler performance

**Comment by bitjammer** [comment]
@swift-ci Please smoke test OS X Platform

**Comment by bitjammer** [comment]
@swift-ci Please test Windows platform

**Comment by bitjammer** [comment]
Looks like some sort of LLDB build failure:

```
Error copying directory from "/Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/buildbot_incremental/swift-macosx-x86_64/lib/swift" to "/Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/buildbot_incremental/lldb-macosx-x86_64/bin/LLDB.framework/Versions/A/Resources/Swift".
```

Will check back tomorrow.

**Comment by bitjammer** [comment]
@swift-ci Please test Windows platform

**Comment by bitjammer** [comment]
@swift-ci Please smoke test OS X Platform

**Comment by bitjammer** [comment]
@swift-ci Please smoke test

**Comment by bitjammer** [comment]
@swift-ci Please test

**Comment by bitjammer** [comment]
@swift-ci Please test Windows platform

**Comment by bitjammer** [comment]
@swift-ci Please test

**Comment by bitjammer** [comment]
@swift-ci Please smoke test

**Comment by bitjammer** [comment]
@swift-ci Please test Windows platform

**Comment by bitjammer** [comment]
@swift-ci Please test

**Comment by bitjammer** [comment]
@swift-ci Please test

**Comment by bitjammer** [comment]
@swift-ci Please smoke test

**Comment by bitjammer** [comment]
@swift-ci Please test Windows platform

**Comment by bitjammer** [comment]
@swift-ci Please test Windows platform

**Comment by bitjammer** [comment]
@nkcsgexi Can you please review the `.swiftsourceinfo` changes?

**Comment by bitjammer** [comment]
Ping. Can I just get a quick look at this please?

File: lib/Serialization/SourceInfoFormat.h [hunk 1]
```diff
@@ -41,15 +41,15 @@ const unsigned char SWIFTSOURCEINFO_SIGNATURE[] = { 0xF0, 0x9F, 0x8F, 0x8E };
...
+const uint16_t SWIFTSOURCEINFO_VERSION_MAJOR = 2;
 
 /// Serialized swiftsourceinfo format minor version number.
 ///
 /// Increment this value when making a backwards-compatible change that might be
 /// interesting to test for. A backwards-compatible change is one where an \e
 /// old compiler can read the new format without any problems (usually by
 /// ignoring new information).
-const uint16_t SWIFTSOURCEINFO_VERSION_MINOR = 1; // Last change: skipping 0 for testing purposes
+const uint16_t SWIFTSOURCEINFO_VERSION_MINOR = 0; // Last change: add doc comment ranges
```
Comment (nkcsgexi) [inline lib/Serialization/SourceInfoFormat.h]: Could you use 2 here instead? we should only increment the version number.

**Review by nkcsgexi — APPROVED** [review]
LGTM! Just a minor comment.

File: lib/Serialization/SourceInfoFormat.h (hunk 1 above)
Comment (bitjammer) [reply lib/Serialization/SourceInfoFormat.h]: Sure thing. I should double-check then, I was following this:

```
/// Increment this value when making a backwards-incompatible change, i.e. where
/// an \e old compiler will \e not be able to read the new format. This should
/// be rare. When incrementing this value, reset SWIFTSOURCEINFO_VERSION_MINOR to 0.
```

I thought that, changing the size of the `DeclLocationsTableData` meant that we had to increase the major version in this case, resetting the minor version. Maybe that's not necessary though?

**Comment by bitjammer** [comment]
Thanks so much! Quick question above based on your comment.

**Pull request merged**